lib/compress.js

```diff
@@ -4499,13 +4501,38 @@ merge(Compressor.prototype, {
     });
 
     OPT(AST_Arrow, function(self, compressor){
-        if (self.body.length === 1 && self.body[0] instanceof AST_Return) {
+        if (!(self.body instanceof AST_Node)) tighten_body(self.body, compressor);
+        if (compressor.option("arrows")
+            && self.body.length == 1
+            && self.body[0] instanceof AST_Return) {
             var value = self.body[0].value;
             self.body = value ? value : [];
         }
         return self;
     });
 
+    OPT(AST_Function, function(self, compressor){
+        tighten_body(self.body, compressor);
+        if (compressor.option("arrows")
+            && compressor.option("ecma") >= 6
+            && !self.name
+            && !self.async
```

Comment:
Since #2153 was merged `&& !self.async` can be removed. An additional test demonstrating this would be a bonus.